New core setting : shareapi_only_share_with_group_members_exclude_gro…

smarinier opened the pull request:
…up_list (issue #37677)


* Resolves: #37677

## Summary
This implements new settings 'shareapi_only_share_with_group_members_exclude_group_list' wich complements the actual boolean setting 'shareapi_only_share_with_group_members'. Indeed the admin may need to isolate some specific groups (eg admin groups) from this consideration of "we can share each others".
- Implement the new array of groups settings, enabled/hidden in admin view by shareapi_only_share_with_group_members enabled/disabled.
- Implement the complement group filtering in different part of code

### Screenshots

Here is the list of users with their groups.

![users-list-with-groups](https://github.com/nextcloud/server/assets/28636549/b3b1ad30-1e38-42c5-a532-3e4a6310df6a)

In the demonstration, I log in with the **alice**'s account and she is in the "DSI" and "Groupe A" groups.

#### Before

From the Settings > Sharing page.
I checked "Restrict users to only share with users in their groups" in the checkbox.

![settings-sharing-before](https://github.com/nextcloud/server/assets/28636549/b2ad51ac-cfd6-4498-8c5c-3790a26a239b)

As "alice", I would like to share my "Project-A" folder.

[alice-search-users-before.webm](https://github.com/nextcloud/server/assets/28636549/44c35a4d-8e20-4532-a756-e8228ea767a7)

#### After

But, I would like, as "alice", to list the users who are in the same group as me to share my folder... Except for some groups where I wouldn't want to list those users.

So, from the Settings > Sharing page.
I checked "Restrict users to only share with users in their groups" in the checkbox... and a search bar appears.
I enter the groups I don't want to see in the shares.

In the example, I enter the "DSI" and "RH" groups.

![settings-sharing-after](https://github.com/nextcloud/server/assets/28636549/a6cc6988-373d-4ff9-adc2-0b021d8e8bf1)

As "alice", I would like to share my "Project-A" folder.

[alice-search-users-after.webm](https://github.com/nextcloud/server/assets/28636549/1ee56310-064c-46ef-978e-3071ad4851b0)

And now, I cannot see the user "john" because he is in the DSI group. And yet, we are in the same group.

## TODO

- [ ] Should be handled in "dav" application, to be really picky
   * apps/dav/lib/Command/MoveCalendar.php : aff
   * apps/dav/lib/Connector/Sabre/Principal.php :
   *apps/dav/lib/DAV/GroupPrincipalBackend.php 

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

artonge commented:
Hey, thanks for the pull request. After a quick glance, it seems to look good, but before diving into it further:

- [ ] DCO does not pass
- [ ] Can you add some screenshots?
- [ ] Can you explain what the change does?
- [ ] You can remove the changes in l10n, this will be handled separately

zak39 commented:
> Hey, thanks for the pull request. After a quick glance, it seems to look good, but before diving into it further:
> 
>     * [ ]  DCO does not pass
> 
>     * [ ]  Can you add some screenshots?
> 
>     * [ ]  Can you explain what the change does?
> 
>     * [ ]  You can remove the changes in l10n, this will be handled separately

Hi @artonge :slightly_smiling_face: 

I edited the first comment to answer you :slightly_smiling_face: 

I think we have achieved our goals ?

If this Pull Request it's okay for you. Do you think it's possible to backport it to the stable25 ?

artonge commented on lib/private/Collaboration/Collaborators/MailPlugin.php:
```diff
@@ -91,6 +93,14 @@ public function __construct(IManager $contactsManager,
+
+		if ($this->shareWithGroupOnly) {
+			$shareWithGroupOnlyExcludeGroups = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members_exclude_group_list', '');
+			$decodedExcludeGroups = json_decode($shareWithGroupOnlyExcludeGroups, true);
+			$this->shareWithGroupOnlyExcludeGroupsList = $decodedExcludeGroups ?? [];
```
```suggestion
			$this->shareWithGroupOnlyExcludeGroupsList = json_decode($shareWithGroupOnlyExcludeGroups, true) ?? [];
```

smarinier replied:
See before

artonge commented on lib/private/Collaboration/Collaborators/UserPlugin.php:
```diff
@@ -96,6 +98,14 @@ public function __construct(IConfig $config,
+
+		if ($this->shareWithGroupOnly) {
+			$shareWithGroupOnlyExcludeGroups = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members_exclude_group_list', '');
+			$decodedExcludeGroups = json_decode($shareWithGroupOnlyExcludeGroups, true);
+			$this->shareWithGroupOnlyExcludeGroupsList = $decodedExcludeGroups ?? [];
```
```suggestion
			$this->shareWithGroupOnlyExcludeGroupsList = json_decode($shareWithGroupOnlyExcludeGroups, true) ?? [];
```

artonge reviewed (commented):
Some nitpicks, but looks fine otherwise

smarinier commented on lib/public/Share/IManager.php:
```diff
@@ -415,6 +415,15 @@ public function shareApiLinkAllowPublicUpload();
+	 * @return array
+	 * @since 27.0.0
+	 */
+	public function shareWithGroupMembersOnlyExcludedGroupsList();
```
Hi, I take example from other "GroupsList" in other places in the code. And i think the "d" from exclude means something. The groups within this "Group(s)List" are excluded from the "Same Group Members" comparaison. (See Later).

smarinier commented on lib/public/Share/IManager.php:
```diff
@@ -415,6 +415,15 @@ public function shareApiLinkAllowPublicUpload();
 
+	/**
+	 * If shareWithGroupMembersOnly is enabled, return an optional
+	 * list of groups that must be excluded from the principle of
+	 * belonging to the same group.
```
In fact, it does not exclude users. The groups are excluded from "same group comparaison". Eg, User1 is member of G1 and G3, User2 of G2 and G3, and User3 of G1 and G3. If G3 is part of this new settings, it won't be used to determine "same group". So, comparaison will be done only on G1 and G2 (User1 and User3 are "same group" but not User2).

smarinier commented on apps/settings/lib/Settings/Admin/Sharing.php:
```diff
@@ -76,6 +76,9 @@ public function getForm() {
 			? implode('|', json_decode($excludedPasswordGroups, true)) : '';
 
+		$shareWithGroupMembersExcludeGroups = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members_exclude_group_list', '');
+		$shareWithGroupMembersExcludeGroupsList = !is_null(json_decode($shareWithGroupMembersExcludeGroups))
+			? implode('|', json_decode($shareWithGroupMembersExcludeGroups, true)) : '';
```
Yes, there are 3 other groupsList in settings in Admin/Sharing. I based my development on them. They use OC.Settings.setupGroupsSelect() that seems to be a jQuery component used in the settings views, when there is a group selection.

come-nc commented:
You will have to be really careful with the wording, because it was not clear to me at all before seeing the screencasts.

I thought this was about allowing some users to share with everyone (not be restricted).

But this is the other way around, it’s about restricting even more.
Can you explain the usecase here, why is it so important to hide some of these groups from sharing option?

come-nc commented on apps/settings/templates/settings/admin/sharing.php:
```diff
@@ -185,6 +185,13 @@
+		<p id="selectShareWithGroupMembersExcludeGroups" class="indent <?php if (!$_['onlyShareWithGroupMembers'] || $_['shareAPIEnabled'] === 'no') {
+	p('hidden');
+} ?>">
+			 <em><?php p($l->t('Exclude some groups from sharing with users in their group')); ?></em>
```
People in these groups are still allowed to share stuff, so the wording is unclear I think.

come-nc commented on lib/private/Collaboration/Collaborators/GroupPlugin.php:
```diff
@@ -45,6 +45,8 @@ class GroupPlugin implements ISearchPlugin {
 	protected $shareeEnumerationInGroupOnly;
 	/** @var bool */
 	protected $groupSharingDisabled;
+	/** @var array */
+	protected $shareWithGroupOnlyExcludeGroupsList;
```
```suggestion
	protected array $shareWithGroupOnlyExcludeGroupsList;
```

come-nc commented on lib/private/Collaboration/Collaborators/MailPlugin.php:
```diff
@@ -53,6 +53,8 @@ class MailPlugin implements ISearchPlugin {
 	protected $shareeEnumerationFullMatch;
 	/* @var bool */
 	protected $shareeEnumerationFullMatchEmail;
+	/** @var array */
+	protected $shareWithGroupOnlyExcludeGroupsList;
```
```suggestion
	protected array $shareWithGroupOnlyExcludeGroupsList;
```

come-nc commented on lib/private/Collaboration/Collaborators/UserPlugin.php:
```diff
@@ -60,6 +60,8 @@ class UserPlugin implements ISearchPlugin {
 	protected $shareeEnumerationFullMatchEmail;
 	/* @var bool */
 	protected $shareeEnumerationFullMatchIgnoreSecondDisplayName;
+	/** @var array */
+	protected $shareWithGroupOnlyExcludeGroupsList;
```
```suggestion
	protected array $shareWithGroupOnlyExcludeGroupsList;
```

come-nc commented on lib/private/Collaboration/Collaborators/UserPlugin.php:
```diff
@@ -60,6 +60,8 @@ class UserPlugin implements ISearchPlugin {
 	protected $shareeEnumerationFullMatchEmail;
 	/* @var bool */
 	protected $shareeEnumerationFullMatchIgnoreSecondDisplayName;
+	/** @var array */
+	protected $shareWithGroupOnlyExcludeGroupsList;
```
```suggestion
	protected array $shareWithGroupOnlyExcludeGroupsList;
```

come-nc replied:
Or even
```suggestion
	/** @var string[] */
	protected array $shareWithGroupOnlyExcludeGroupsList;
```

come-nc commented on lib/private/Share20/Manager.php:
```diff
@@ -1940,6 +1948,22 @@ public function shareWithGroupMembersOnly() {
+		if (!$this->shareWithGroupMembersOnly()) {
+			return [];
+		}
+		$excludeGroups = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members_exclude_group_list', '');
+		$decodedExcludeGroups = json_decode($excludeGroups, true);
+		return $decodedExcludeGroups ?? [];
```
This logic appears at least 5 times in this PR, may be the time for adding a `getAppValueArray` method in IConfig.

smarinier replied:
There is no getAppValueBool, getAppValueInt, aso... I think it would be disturbing to have only this very specific method in an interface. But i think you're right, if you consider all the getAppValue(...) == 'yes', there is a need on getting settings. But maybe it should be handled in a dedicated PR ?

zak39 commented:
Hi @come-nc :slightly_smiling_face: 

In fact, we wish to add an additional option to the Sharing settings available for the instance admins (Settings > Administration > Sharing).
 
One of the current available option is "Restrict users to only share with users from their groups"
 
If the box for this setting is checked, we would like to add a selection field + text below this setting:
 
- the text explains what the selection field does:
- we have suggested a text that can be changed later on, so that it can be as clear as possible
- the purpose of this field is to make an exception to the rule "Restrict users to only share with users in their groups" if it is checked:
-> if a group (or groups) is inserted in this field, e.g. the group "Guests", then users in the Guest group will not be able to share with each other
- the selection field will display the user groups of the instance (local and LDAP) and will allow admins to insert one or more groups in this field
 
We believe this concept would be beneficial to the administration of Nextcloud. In this example, the organization could check the setting "Restrict users to only share with users in their groups" and use a “Guest” group in their LDAP to group all their external users. This LDAP Guest group would then be created in Nextcloud, meaning all the guests would be able to see each other, even though they are not from the same companies. **This represents a loss of privacy**.
 
And this is only one example among many others.


One more, the issue linked to this PR has been reviewed @nimishavijay and it was she who proposed the wording in this issue : https://github.com/nextcloud/server/issues/37677 .

zak39 commented on apps/settings/templates/settings/admin/sharing.php:
```diff
@@ -185,6 +185,13 @@
 } ?> />
 			<label for="onlyShareWithGroupMembers"><?php p($l->t('Restrict users to only share with users in their groups'));?></label><br/>
 		</p>
+		<p id="selectShareWithGroupMembersExcludeGroups" class="indent <?php if (!$_['onlyShareWithGroupMembers'] || $_['shareAPIEnabled'] === 'no') {
```
@come-nc Do you think we can write this code to fix the Psalm ?

`if (!$_['onlyShareWithGroupMembers'] || $_['shareAPIEnabled'] ?? 'no' === 'no')`

zak39 commented:
I updated the screenshot where we can see I added the DSI and RH groups in the first comment

nimishavijay commented:
> You will have to be really careful with the wording, because it was not clear to me at all before seeing the screencasts.

The wording is quite tricky with this one, agreed there. Alternative wording:
- [ ] Limit users to share with only users their group
  Restrict some groups from sharing with users in their group
    [ Select groups ]

@jancborchardt or @szaimen for help? 

The use case is quite specific but it seems to make sense to me. Another idea for managing permissions for groups is from the user management page, by having a settings for each group. This proposal seems like a quick fix.

come-nc commented:
> * -> if a group (or groups) is inserted in this field, e.g. the group "Guests", then users in the Guest group will not be able to share with each other

Then again this is not true, they will be able to share with each other but only if they are both members of another group.
This is what is worrying me, it is so hard to describe this feature with words, and that means that whatever we do this is something that users will misunderstand, misuse, and see as a magic box that does *stuff*.

> We believe this concept would be beneficial to the administration of Nextcloud. In this example, the organization could check the setting "Restrict users to only share with users in their groups" and use a “Guest” group in their LDAP to group all their external users. This LDAP Guest group would then be created in Nextcloud, meaning all the guests would be able to see each other, even though they are not from the same companies. **This represents a loss of privacy**.
> 
> And this is only one example among many others.

I understand this example and can see cases where you even have some kind of `users` group with everyone.
But I feel like restricting sharing to groups is already a corner case and not the most common.

What about using "ignore" rather than "exclude" in the wording? "Ignore these groups for the sharing restriction" or something.

And by the way, wording can be fixed later but the variable name also needs to be clear so that we understand the code when we read it.

jancborchardt commented:
Further wording adjustment, based on @nimishavijay’s suggestion. Note we want to switch over from the word "users" to either "people" or "accounts", whichever fits better. (In this case it’s "accounts" as it’s technical.)

> **Limit accounts to share only within their groups**
> Members of selected groups will be restricted to share only with people in the same group
> [ Select groups ]

Does that explain it well?

jancborchardt reviewed (commented):
Further wording adjustment, based on @nimishavijay’s suggestion. Note we want to switch over from the word "users" to either "people" or "accounts", whichever fits better. (In this case it’s "accounts" as it’s technical.)

> **Limit accounts to share only within their groups**
> Members of selected groups will be restricted to share only with people in the same group
> [ Select groups ]

Does that explain it well?

come-nc commented:
> Further wording adjustment, based on @nimishavijay’s suggestion. Note we want to switch over from the word "users" to either "people" or "accounts", whichever fits better. (In this case it’s "accounts" as it’s technical.)
> 
> > **Limit accounts to share only within their groups**
> > Members of selected groups will be restricted to share only with people in the same group
> > [ Select groups ]
> 
> Does that explain it well?

No, this is not what the new option does.
What it does is ignore some of the groups when checking if sharing is allowed.

So if you are in groups A, B and C.
The checkbox is checked and B is listed in the new options.
You will be able to share only with users from A and/or C (which may also be members of B but that is not relevant).

jancborchardt commented:
Hm ok, in that case I am questioning the use-case of this feature.

If the main case as described in the issue is to prevent guests from sharing among each other, it should not be so generic, and just be called:

> **Disallow sharing among guests**
> [Select guest group name]

(Possibly with a good default set)

Also cc @AndyScherzinger @schiessle regarding Files planning as this seems yet another complex feature addition for sharing.

AndyScherzinger commented:
For us to act on this within our plannings we would need to fully understand the use case and agree on having it added.
Else there is not much to do from my side at least.

maximelehericy commented:
@jancborchardt 

I think the guest group is one example.
But in my opinion, there are other examples:
For instance, you use functional groups to allow people to collaborate together (project or cust A, project or cust B), but also technical groups to manage your instance (e.g : people with the right to use Talk, people with the right to use Office, and so on).
When you tick the box Restrict users to share only with members of their groups, you don't want user from cust A to share with user from cust B through the "Talk people" group.
Therefore, you'd like to ignore these "technical" groups.

dorianne-arawa commented:
Hello,
As a colleague of @zak39 , I take the liberty of answering on this subject.

Thank you for all the  feedback! Here are our answers on the different questions raised:

Regarding the wording,  our proposal lacked clarity. Wee think the proposal of @come-nc  is more relevant:
* `Restrict users to only share with users in their groups`
* `Ignore these groups for the sharing restriction`
The correlation between the two sentences is clearer thanks to the words `ignore` and `restriction`.

About the use case : many companies using Nextcloud can be notable providers for various customers (like consulting firms ). If a company uses its Nextcloud with  different customers, they should not be able to see each other for privacy reasons (especially in the field of shares).
However, for management or access rights reasons, it can be convenient for administrators of a Nextcloud platform to define "functional" or "management" groups. These groups gather users and allow to manipulate several of them at once. However, these "functional" or "management" groups are not intended for users to see each other. This can  sometimes be prohibitive to the use of the Nextcloud service.

Concrete examples :
* to make a groupfolder containing all "clients" ("Clients" group)
* for mass communication purposes via sharing (sharing documentation, instructions, administrative documents, etc.) This prevents making lots of shares with the same rights, on the same folder (one share per client, if we have to share with 30 clients or more, that's a lot of shares to manage manually)
* in the Workspace application, the default group "WorkspacesManagers" can gather users of several customers 
* to have a users group(s) allowed to give access to "administrator privileges" of one or more Nextcloud settings sections
* to restrict the application access to a given group of users (application access limitation option in the application management interface)

Even if these example cases might not seem common at first , there remains an existing use case which needs to be addressed to allow a more flexible use of Nextcloud, adapted to different contexts, and as secure and confidential as possible.


Regarding Nimisha's proposal : `Another idea for managing permissions for groups is from the user management page, by having a settings for each group. This proposal seems like a quick fix.`
We think we should refocus on the notion of shares which is addressed here, and which is not managed (at least currently) in the Nextcloud user interface. As a reminder, this feature is meant to keep confidentiality of personal user’s data (names, email address and company), which is visible in the "shares" field.

I hope that my answer and my different arguments will be clear enough, I will answer your questions if there are any misunderstandings.

jancborchardt commented:
Thank you for the clarification @dorianne-arawa! :) That now makes more sense.

@AndyScherzinger @schiessle what do you think about the explanation in https://github.com/nextcloud/server/pull/38173#issuecomment-1551094957 ?

smarinier commented on lib/private/Collaboration/Collaborators/MailPlugin.php:
```diff
@@ -91,6 +93,14 @@ public function __construct(IManager $contactsManager,
+
+		if ($this->shareWithGroupOnly) {
+			$shareWithGroupOnlyExcludeGroups = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members_exclude_group_list', '');
+			$decodedExcludeGroups = json_decode($shareWithGroupOnlyExcludeGroups, true);
+			$this->shareWithGroupOnlyExcludeGroupsList = $decodedExcludeGroups ?? [];
```
Changed for (Json need to be decoded):

			$this->shareWithGroupOnlyExcludeGroupsList = json_decode($this->config->getAppValue('core', 'shareapi_only_share_with_group_members_exclude_group_list', ''), true) ?? [];

smarinier commented:
Hi @jancborchardt , @come-nc and @artonge,

I think we've taken in account all the discussion concerning the wording, and other remarks. Do you think you could have a look on this ?

Thanks a lot

artonge reviewed (commented):
Would it make sense to not bound this feature to the "Restrict users to only share with users in their groups" setting?

dorianne-arawa commented:
Hello @artonge

To answer your question, we believe that both settings are definitely bound ("Restrict users to only share with users in their groups" and "Ignore these groups for sharing restriction").

Let's say the setting 1 is the existing checkbox "Restrict users to only share with users in their groups" and the setting 2 is the new option we would like to add (sentence + multiselect for groups).

The setting 1 is a restriction : it ensures that users no longer see each other (except if they belong to the same group)
The setting 2 offers an additional restriction to setting 1 to prevent users from a same group to see each other (by ignoring specified groups)
As setting 2 is additional to the first one, the 2 are linked.

Arawa

dorianne-arawa commented:
Hi!
To follow up on @jancborchardt 's question, what do you think of our last updates @AndyScherzinger, @artonge and @schiessle ?
Dorianne - Arawa

zak39 commented:
Hi everyone!

Do you have an issue with CI on this PR?

Baptiste - Arawa

AndyScherzinger commented:
@zak39 I enabled the CI (is blocked due to the PR coming from a fork - different remote)

@maximelehericy @artonge is back next week.

zak39 commented:
> @zak39 I enabled the CI (is blocked due to the PR coming from a fork - different remote)
> 
> @maximelehericy @artonge is back next week.

Tanks @AndyScherzinger :)

I guess we have to resolve all the CI's fails ?

AndyScherzinger commented:
> I guess we have to resolve all the CI's fails ?

Can't say for sure, since I am not a php-dev, so I would rely on @artonge's feedback here.

artonge commented on apps/settings/src/admin.js:
```diff
@@ -1,5 +1,5 @@
 window.addEventListener('DOMContentLoaded', () => {
-	$('#linksExcludedGroups,#passwordsExcludedGroups').each(function(index, element) {
+	$('#excludedGroups,#shareGroupMembersExcludeGroups,#linksExcludedGroups,#passwordsExcludedGroups').each(function(index, element) {
```
I cannot find the `excludedGroups` id used anywhere.

smarinier replied:
This is not the final result. This admin.js was modified on on May 16th by 06d8a03315fd72a72229fb48a096e514e3e5b3b0, and excludedGroups was then removed.
As our pull request was started in April, this was fixed after the "merge branch" in June ;). (26cab0629f3b8a5eb6bedf83fc6ba12d230ee734). I don't know why the code review  system does not take this in account.

artonge replied:
The code review system reflect what will be merge. If this change is not intentional, then you need to remove it

smarinier replied:
I guess that. This was already done (June 19th). I need to check what happened to my commit.

smarinier replied:
Solved with last commit

smarinier commented on lib/private/Share20/Manager.php:
```diff
@@ -1931,6 +1939,21 @@ public function shareWithGroupMembersOnly() {
+	 *
+	 * @return array
+	 */
+	public function shareWithGroupMembersOnlyExcludeGroupsList() {
```
I do agree. But on the other side, this function is an "interface function" with the same operating mode as shareWithGroupMembersOnly() (and is very closed and linked with this function). I'm not sure of the purpose of writing a private functions in an interface definition. And if we remove shareWithGroupMembersOnlyExcludeGroupsList() from the interface, it would mean you can write a class with shareWithGroupMembersOnly() handling but not the shareWithGroupMembersOnlyExcludeGroupsList() functionality ?

smarinier commented on apps/settings/templates/settings/admin/sharing.php:
```diff
@@ -187,6 +187,13 @@
+		<p id="selectShareWithGroupMembersExcludeGroups" class="indent <?php if (!$_['onlyShareWithGroupMembers'] || $_['shareAPIEnabled'] === 'no') {
+	p('hidden');
+} ?>">
+			 <em><?php p($l->t('Ignore these groups for the sharing restriction')); ?></em>
```
Ha... I think you should take a look on the long discussion about the wording (previous messages in May). The final chosen sentences came after a somehow collaborative and democratic discussion ;) . Are you sure you want to restart the whole debate (with nimishavijay, come-nc, jancborchardt, AndyScherzinger, schiessle and dorianne-arawa) ? Especially, it was decided not to use "Exclude" that seems ambiguous for the users, but rather use "ignore" and "restriction".

artonge replied:
I did and I feel like the discussion is not over. The scope of the feature is defined, but the wording do not reflect the feature.

nimishavijay replied:
Since this is a fairly technical topic, we could use a main line (in `--color-main-text`) and a subline (in `--color-text-maxcontrast`) that provides further explanation?
Now that we are not restricted to one sentence, we can be more verbose with the wording in the subline. Feel free to correct my wording in case I have misunderstood the feature. It can look something like this:


- [ ] Restrict users to only share with other users in their group
     - [ ] Ignore some groups for the sharing restriction
        The mentioned groups will not be able to share within their own group, and can share with other groups.
        ` [Group1]  [Group2] `

What do you think?

artonge replied:
It is a nice idea to add a subline, but I still think that `Ignore some groups for the sharing restriction` is not reflecting the feature.

artonge replied:
No, because they will still see each other if they are part of another group

nimishavijay replied:
"Disallow some groups from sharing within their own group"?

maximelehericy replied:
In https://github.com/nextcloud/server/pull/38173#issuecomment-1544016290 @come-nc rejected the "exclude" term in that proposition "Exclude some groups from sharing with users in their group" because he thought it was not reflecting the feature. 

So in itself, the word exclude is not banned.

Then, I agree with @artonge's last proposal here https://github.com/nextcloud/server/pull/38173#discussion_r1318756620, because it is exactly what the feature does. It may not be easy to understand, but it is a setting addressed to Nextcloud admins who we can expect to understand. This could also be documented in the admin manual with  longer example.

In addition, with Nimisha's subline proposal, maybe we could rephrase @come-nc example:
> So if you are in groups A, B and C.
> The checkbox is checked and B is listed in the new options.
> You will be able to share only with users from A and/or C (which may also be members of B but that is not relevant).

In something like:
If a user is member of groups A, B and C, and B is in the exluded group list, the user will be able to share only with users from A and/or C (which may also be members of B but that is not relevant).

come-nc replied:
I still think ignore is better, because we are ignoring groups when searching for restrictions, and we are not excluding anything.
But I’m not an english-native speaker so what’s logical to me might not be correct.

dorianne-arawa replied:
Hello,

Following various discussions, we propose the following formulations, in French and English. If you agree, we'll make the necessary changes to the code: 
- French : "Ignorer ces groupes lors de la vérification de l'appartenance des comptes aux groupes" (seen with @come-nc , among other )
- English : "Ignore the following groups when checking group membership" (seen with @jancborchardt , among others)

Have a nice day,
Dorianne - Arawa

smarinier replied:
Hi,
@artonge , @come-nc,  @jancborchardt 
@dorianne-arawa
(and others)

I've commited this label proposal (and fixed the last merge conflict that have recently arisen ). Please note that the french part of this message is not handled (at this moment) as required at the beginning of the PR approval process.

Have a nice WE,
Sébastien - Arawa

zak39 commented:
Hi everyone :)

I don't understand why the dco is failled ? Can you explain me, please ?

I saw all commits have sign-off

dorianne-arawa commented:
Indeed, we have heard @artonge 's comments and are working on the subject to propose a new wording and submit it to you.

AndyScherzinger commented:
> I don't understand why the dco is failled ? Can you explain me, please ?
>
> I saw all commits have sign-off

@zak39 see at the very end when opening the dco check:

> Summary

Commit sha: [b1f6ba2](https://github.com/nextcloud/server/pull/38173/commits/b1f6ba2e0e4bee312a588363f8c4fed350d1767b), Author: zak39, Committer: zak39; Expected "zak39 [fotia.baptiste@hotmail.com](mailto:fotia.baptiste@hotmail.com)", but got "Baptiste Fotia [fotia.baptiste@hotmail.com](mailto:fotia.baptiste@hotmail.com)".

zak39 commented:
> > I don't understand why the dco is failled ? Can you explain me, please ?
> > I saw all commits have sign-off
> 
> @zak39 see at the very end when opening the dco check:
> 
> > Summary
> > Commit sha: [b1f6ba2](https://github.com/nextcloud/server/pull/38173/commits/b1f6ba2e0e4bee312a588363f8c4fed350d1767b), Author: zak39, Committer: zak39; Expected "zak39 [fotia.baptiste@hotmail.com](mailto:fotia.baptiste@hotmail.com)", but got "Baptiste Fotia [fotia.baptiste@hotmail.com](mailto:fotia.baptiste@hotmail.com)".

But it's strange ! My sing-off is present :monocle_face: 

So, maybe  I have to add a blank line or a message ? :thinking: 

I will try it.

botsarenthuman commented:
Hi, is there any movement on this?

smarinier commented:
> Hi, is there any movement on this?

The code has been ok for a few months. But there were discussions about the message labels. We are waiting now for PR approvals.

artonge commented:
@smarinier I think your rebase went wrong:

![image](https://github.com/nextcloud/server/assets/6653109/d9fd0c21-ca64-439e-919e-0b285903f327)

smarinier commented:
> @smarinier I think your rebase went wrong:

We have DCO issues. Because of signing off. The command given by the DCO help comments doesn't work. Could we have git help on this stuff. I wouldn't like to put more garbage in the code than the fact the commit comment does not have exactly the expected name.

botsarenthuman commented:
This is exciting, will we be able to add groups via `occ` ?

zak39 commented:
Hi :wave:

We have a few conflicts in the assets (`dist/*`) and I think we have them because you accept Pull-Requests for the `stable28.0.0-beta1` or `stable28.0.0-beta1` branch.

Should we rebase this branch and build assets again ?

AndyScherzinger commented:
we haven't branches-off v28 yet, so v28 is still bound to the `master` branch until we hit RC1 and create `stable28`

artonge commented:
> Should we rebase this branch and build assets again ?

The answer is yes, but you can wait for the second approval before doing the last rebase ;).

zak39 commented:
> we haven't branches-off v28 yet, so v28 is still bound to the `master` branch until we hit RC1 and create `stable28`

I'm sorry, I misread this PR and I thought you have a `stable28.0.0beta1` branch ^^'

https://github.com/nextcloud/server/pull/41228


> > Should we rebase this branch and build assets again ?
> 
> The answer is yes, but you can wait for the second approval before doing the last rebase ;).

Good ! :+1: 
We are waiting the second approval in that case :wink:

artonge commented:
Need update for the front-end part as the share settings have been migrated : https://github.com/nextcloud/server/pull/41581/files#diff-387668c3bef3964472fa50bcc5b25c63c2534e018af9af718114430f5eb19299

zak39 commented:
> Need update for the front-end part as the share settings have been migrated : https://github.com/nextcloud/server/pull/41581/files#diff-387668c3bef3964472fa50bcc5b25c63c2534e018af9af718114430f5eb19299

Hi @artonge :wave: 

So, I guess we have to rebase our branch to get the new component ?
Should we to build assets too ?

AndyScherzinger commented:
master-branch moved to 29 since we tagged RC1 of 28 on Thursday. Currently both branches master/stable28 are mostly the same. So currently you'd be very likely compatible with both.

AndyScherzinger commented:
> Is it okay for you?
> 
> Do we need to make any additional changes?

I lack the tech knowledge, so I'll depend on @artonge's feedback on that

artonge commented:
No, this looks good.

@come-nc @susnux, @Fenn-CS, this need a review.

dorianne-arawa commented:
Hi @come-nc @susnux and @Fenn-CS !

Is there any news about this PR  ?
We are open to feedbacks :)

susnux reviewed (approved):
Looks good but we definitely need a section in the admin docs to make clear what is it about.

Also the PR needs a rebase and compile. Maybe squash the commits?

artonge commented:
/compile /

zak39 commented:
> /compile /

Nextcloud bot could not build as far as I can see.

Do you want to build assets manually ?

artonge commented:
> Do you want to build assets manually ?

Please do, and also squash your commits

zak39 commented:
> > Do you want to build assets manually ?
> 
> Please do, and also squash your commits

I will build assets this morning.
For the squash, if I understand, we have to squash (merge commits) and get a few commits ?

What is the good practice for this command ? Squash all "Merge branch" commits ? Squash all feature commits ?

artonge commented:
> What is the good practice for this command ? Squash all "Merge branch" commits ? Squash all feature commits ?

This is only to have a cleaner git history. You can try:

```sh
git reset --soft HEAD~3
git commit
```

Where 3 must be replaced by the number of commits you want to squash. In your case, probably 18.

zak39 commented:
> > What is the good practice for this command ? Squash all "Merge branch" commits ? Squash all feature commits ?
> 
> This is only to have a cleaner git history. You can try:
> 
> ```shell
> git reset --soft HEAD~3
> git commit
> ```
> 
> Where 3 must be replaced by the number of commits you want to squash. In your case, probably 18.

I don't think it's a better solution to use the `git reset` command. Because, I see files we didn't touch.
Application files in the `apps/` folder. I'm trying with squash from the `git rebase` command.

However, I have to get to `5af683d2c4` commit to squash our first commit.

zak39 committed:
feat(php,vue): Create a new parameter

Signed-off-by: Baptiste Fotia <fotia.baptiste@hotmail.com>

zak39 commented:
@artonge it's okay now :+1:

zak39 commented:
Thank you a lot @skjnldsv for your help :pray:

artonge commented:
@zak39 Node CI is still failing.

zak39 committed:
test(php): Fix the unit tests

I added our context in the unit tests following the advice of Louis.

Link : https://github.com/nextcloud/server/pull/43186#issuecomment-1916571027

Signed-off-by: Baptiste Fotia <fotia.baptiste@hotmail.com>

zak39 committed:
build(): build assets

Signed-off-by: Baptiste Fotia <fotia.baptiste@hotmail.com>

artonge commented:
Drone CI is waiting as this is a fork PR, but it is green in this copy https://github.com/nextcloud/server/pull/43186, merging.

zak39 commented:
> @zak39 Have you properly rebased? Looks like there is some merge conflicts left.

It's strange ! I rebased correctly yesterday. I try rebase again now to see.

zak39 committed:
build(): build assets

Signed-off-by: Baptiste Fotia <fotia.baptiste@hotmail.com>

artonge commented:
Drone CI is waiting as this is a PR from a fork, but it is green in this PR: https://github.com/nextcloud/server/pull/43186
So merging.

zak39 commented:
Oooh good, thanks @artonge :pray: 

Let's keep our an appointment or it's not necessary ?
